doc/dev/error-categorization.md

```diff
@@ -0,0 +1,167 @@
+and developers simply cannot know each and every code.
+
+To get an idea of programmatic reactions take a method call which returns a `Timeout` error. Naturally
+a senseful reaction would be to retry at a later point in time. So
+the reaction here would be to retry with a time based approach. On the other hand
+it does not make sense to differentiate between `No Write Permission` and `No Read Permission` as
+the application just knows that it simply cannot access the desired resource and tells
+the user to grant it.
+
+Categories are leaf based, so you cannot put an error
+into a node (branch) such as `Permanent errors` (see below). If you feel for a new category,
+please forge a design decision document and make a PR to Elektra's repo.
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an error/warning in there.
+
+The current structure looks like this:
+
+- Permanent errors
+  - Resource
+    - Memory Allocation
+    - General Resource
+  - Installation
+  - Logical
+    - Assertion
+    - Interface
+    - Broken plugin
+- Conflicting State
+- Timeout
+- Out-Of-Range
+- Validation
+  - Syntactic
+  - Semantic
+
+### Permanent errors ("C01000")
+
+The branch category `Permanent Errors` refer to such errors which cannot be fixed by retry
+at all. `Permanent Errors` are subdivided into
+
+- Resource
+- Installation
+- Logical
+
+#### Resource ("C01100")
+
+`Resource Errors` as a branch category are all kinds of errors which are either permission related, existence related
+or missing resources such as memory, RAM, etc.
+Examples are missing files, insufficient permissions or out of memory.
+This category forces users or applications to provide additional resources, create a missing file/ directory etc
+or provide the relevant permissions for Elektra to work.
+
+##### Memory Allocation ("C01110")
+
+`Memory Allocation Errors` are errors which come from failed `malloc` calls primarily as no
+more memory could be allocated for the application. Such errors will gain special handling
+in future releases.
+
+##### General Resource ("C01120")
+
+`General Resource Errors` are all kind of permission and existence errors. Examples are
+missing files/ directories or insufficient permission to execute certain commands (eg. you
+would require sudo permissions). Reactions are fixing the permissions or creating the file/directory
+and retry the operation.
+
+#### Installation ("C01200")
+
+`Installation Errors` are those errors which are related to a wrong installation such as
+wrong plugin names, missing backends, initialization errors, misconfiguration of Elektra etc.
+Installation errors might also be non-Elektra specific but also from dependent library/applications
+such as gpg.
+Also plugin configuration errors belong to `Installation Errors` as they this happens during
+mounting.
+Users will have to reconfigure, reinstall, recompile (with other settings) Elektra in order to
```

Comment:
Very good explanation! Actually you can make this structure everywhere within this document:

1. describe which errors belong the the category
2. which errors do not belong there
3. comparison with other categories
4. examples of this error
5. what users can do about it

Reply:
I tried to apply this form to all leaf categories even though 2 and 3 are very difficult to apply to all categories.